### PR TITLE
RELATED: BB-1502 Fix PivotTable offset paging issue

### PR DESCRIPTION
**Pull request by martinmosko**
---

[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist

- [x] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [x] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [x] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [x] Migration guide (for major changes) is mentioned in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [ ] Successful `extended test - examples`
- [x] Successful `extended test - storybook`
- [x] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)

# Related PRs

- gdc-app-components:
- gdc-analytical-designer:
- gdc-dashboards:

# Related Jira tasks

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3511/ : SUCCESS in 7m 08s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4353/ : SUCCESS in 15s

File: src/components/core/pivotTable/agGridDataSource.ts [hunk 1]
```diff
@@ -30,6 +30,15 @@ export const areTotalsChanged = (gridApi: GridApi, newTotals: IGridTotalsRow[])
     return false;
 };
 
+const isInvalidGetRowsRequest = (startRow: number, gridApi: GridApi): boolean => {
+    const bottomRowIndex = ApiWrapper.getPaginationBottomRowIndex(gridApi);
+    if (typeof bottomRowIndex === "number") {
```
Comment (strouhaljan) [inline src/components/core/pivotTable/agGridDataSource.ts]: How about `if (bottomRowIndex !== null)`? `null` is return value of `ApiWrapper.getPaginationBottomRowIndex()` saying that something is wrong.

File: src/components/core/pivotTable/tests/agGridDataSource.spec.ts [hunk 2]
```diff
@@ -8,19 +8,22 @@ import { createIntlMock } from "../../../visualizations/utils/intlUtils";
 const intl = createIntlMock();
 
 describe("getGridDataSource", () => {
-    function createMockedDataSource() {
+    function createMockedDataSource(customGridApi?: () => any) {
```
Comment (strouhaljan) [inline src/components/core/pivotTable/tests/agGridDataSource.spec.ts]: How about renaming `customGridApi` to `customGridApiProps` and use spread operator while creating mocked grid API? See below.

File: src/components/core/pivotTable/tests/agGridDataSource.spec.ts [hunk 3]
```diff
@@ -8,19 +8,22 @@ import { createIntlMock } from "../../../visualizations/utils/intlUtils";
...
         const cancelPagePromises = jest.fn();
         const onSuccess = jest.fn();
-        const getGridApi = () => ({
-            setPinnedBottomRowData: jest.fn(),
-            getPinnedBottomRowCount: jest.fn(),
-            paginationProxy: {
-                bottomRowIndex: 32,
-            },
-        });
+        const getGridApi = customGridApi
```
Comment (strouhaljan) [inline src/components/core/pivotTable/tests/agGridDataSource.spec.ts]: See above:
```
    const getGridApi = () => ({
        setPinnedBottomRowData: jest.fn(),
        getPinnedBottomRowCount: jest.fn(),
        paginationProxy: {
            bottomRowIndex: 32,
        },
        ...customGridApiProps,
    });
```

File: src/components/core/pivotTable/tests/agGridDataSource.spec.ts [hunk 4]
```diff
@@ -70,19 +74,45 @@ describe("getGridDataSource", () => {
...
+
+    it("should ignore getRowsRequest validation when gridApi is not providing expected data", async () => {
+        const resultSpec = pivotTableWithColumnAndRowAttributes.executionRequest.resultSpec;
+        const {
+            getPage,
+            successCallback,
+            gridDataSource,
+            onSuccess,
+            cancelPagePromises,
+        } = createMockedDataSource(() => ({
```
Comment (strouhaljan) [inline src/components/core/pivotTable/tests/agGridDataSource.spec.ts]: See above:
```
} = createMockedDataSource({
    paginationProxy: {},
});
```

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3519/ : SUCCESS in 6m 41s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4361/ : SUCCESS in 13s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3521/ : SUCCESS in 4m 57s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4363/ : SUCCESS in 10s

**Comment by martinmosko** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1875/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1218/ : SUCCESS in 12m 28s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4366/ : SUCCESS in 10s

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- gooddata-react-components-storybook-zuul-docker : SKIPPED
- gooddata-react-components-unit-tests-zuul-docker-el7 : SKIPPED
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/961/ : FAILURE in 24s
- gooddata-react-components-examples-zuul-docker : SKIPPED

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1881/ : SUCCESS in 17m 03s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3529/ : SUCCESS in 9m 34s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/962/ : SUCCESS in 17s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/1038/ : FAILURE in 36m 36s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1884/ : SUCCESS in 15m 20s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3532/ : SUCCESS in 3m 22s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/965/ : SUCCESS in 55s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/1042/ : SUCCESS in 33m 15s

**Pull request merged**